Wip/detach v10

rellla opened the pull request:
Though this should be kept as a draft until it's tested enough, it's ready for review.
Commit 1-3 do some refactoring of the classes first
Commit 4 does a reordering in start/stop mechanism
Commit 5 removes cVideoRender as a member of cVideoDecoder... we have a code duplication, which could be done better. At best with introducing a cFrame class...
Commit 6 moves plugin start and stop to the state machine. The plugin starts in attached mode first - which does the init. When the plugin stops, it enters detached state before the real exit - which is a preparation for the next commit.
Commit 7 introduces detach/attach via svdrp

This is only basically tested on rpi4 with start/stop and detach/attach. Needs testing to detach in all other states.

rellla committed:
config: SetupParse only does parsing

It does not set any values but let the other objects fetch the values.

rellla committed:
device: refactor plugin start and stop

rellla committed:
drop ConfigMakePrimary because it isn't used anyway

rellla committed:
move decoding thread to videostream

rellla committed:
codec_video: no need to have render as a member

TODO: IsInterlacedFrame() is duplicated

rellla committed:
start/stop: use state machine

rellla committed:
add detach/attach via svdrp

fwolter commented on softhddevice-drm-gles.cpp:
```diff
@@ -288,6 +283,22 @@ cString cPluginSoftHdDevice::SVDRPCommand(const char *command,
+			return "SoftHdDevice is already detached";
+
+		m_pDevice->Detach();
+		return "SoftHdDevice was detached";
```
Maybe use active:
```suggestion
		return "Detached SoftHdDevice";
```

fwolter commented on softhddevice.cpp:
```diff
@@ -323,95 +292,137 @@ void cSoftHdDevice::HandlePause(void)
 	switch (m_state) {
+		case State::DETACHED:
+			std::visit(overload{
+				[this, &needsResume, &invalid](const PlayEvent&) { invalid(); },
```
Only the values used inside the lambda needs to be specified in the captor arguments.
```suggestion
				[&invalid](const PlayEvent&) { invalid(); },
```

fwolter commented on softhddevice.cpp:
```diff
@@ -453,6 +464,52 @@ void cSoftHdDevice::OnEnteringState(enum State state) {
+			m_audioReassemblyBuffer.Reset();
+
+			// now do the detach
+			m_pVideoStream->DecodingThreadResume();
```
Maybe move this to ExitDisplayThread() etc.. That the threads need to be resumed when they are about to stop is an implementation detail and looks a bit odd from the outside.

rellla replied:
Would that make it better?
I'm not satisfied with the halt/resume no halt/no resume mechanism in OnEventReceived either. Maybe we find some nice refactoring here?

fwolter replied:
When moving to synchronized cQueue we need to rethink the whole resume/halt mechanism. Maybe lets keep it for now and refactor it when implementing sync cQueue.

fwolter commented on softhddevice.cpp:
```diff
@@ -844,6 +905,9 @@ int cSoftHdDevice::PlayAudio(const uchar *data, int size, uchar id)
 {
 //	LOGDEBUG("device: %s: %p %p %d %d", __FUNCTION__, this, data, size, id);
 
+	if (m_skipstream)
```
You could check the state instead. This can be done without synchronization, because the state is only changed by the VDR main thread, and Play*() are only called by it.
```suggestion
	if (m_state == DETACHED)
```

rellla replied:
I had that one already, but it didn't work, because then the state gets "ATTACHED" before the things in OnEnteringState(ATTACHED). So VDR already already sends data without having cSoftHdAudio ... that ended in a segfault.
Maybe we should change SetSate to
```
if (m_state != newState) {
		LOGDEBUG("device: Preparing to leave state %s", StateToString(m_state));
		OnLeavingState(m_state);
		LOGDEBUG("device: Changing state %s -> %s", StateToString(m_state), StateToString(newState));
		OnEnteringState(newState);
		m_state = newState;
		LOGDEBUG("device: State changed to %s", StateToString(m_state));
	}
```
?

fwolter replied:
Right of the bat, I think that makes more sense generally.

fwolter replied:
Actually the order shouldn't matter if all state changes and Play*() are processed in the VDR main thread. I think my assumption is wrong that Play*() are called from the main thread. In this case there needs to be a mutex around the state and everywhere where the state is accessed.

rellla replied:
I guess we can assume that Play* is called from the VDR main thread or player thread. I don't think we need to lock anything.
I'll try with the above code. It follows the logic of the state change as it sets m_state not until we really reached it.

rellla replied:
Changing that will follow the logic but we open the next problem:
When entering DETACHED, the first we want it to skip data in Play* - But DETACHED is set not until OnEnteringState has finished then. So, though it works for entering ATTACHED it doesn't work for entering DETACHED anymore. I'd suggest to use m_skipStream for now.

rellla replied:
So i have to correct myself. We need to lock the state machine, because when we use svdrp (which is a separate thread) we run into a problem if we are in between a state change and want to set a new state....

rellla replied:
https://github.com/rellla/vdr-plugin-softhddevice-drm-gles/pull/113/commits/3230c8c1da537b1c4c07ffe112792955d81403c8 should fix that

fwolter replied:
Since there are three threads, now, changing the state (main, player, svdrp), we need the mutex here, too. Otherwise, the decoder could be closed, while we are in Play*().

rellla replied:
You mean locking the whole PlayVideo and PlayAudio to protect the m_skipStream?
I guess, we can use the same mutex for both, because PlayAudio and PlayVideo are called from the same VDR thread?

We have to look into PlayVideoPkts/PlayAudioPkts afterwards, though.

fwolter replied:
Not only to protect m_skipStream, but also to protect the logic inside the decoders. Otherwise a thread can change/close the decoder while another thread sends data to it via Play*(). I'd say we *must* use the same mutex for both play methods (as for the state mutex).

rellla replied:
Let me push sth...

fwolter commented on videostream.cpp:
```diff
@@ -273,3 +274,39 @@ void cVideoStream::Open(AVCodecID codecId, AVCodecParameters *par, AVRational ti
+ * @retval     1 if active
+ *             0 if stopped
+ */
+bool cVideoStream::DecodingThreadIsActive(void) {
```
If you put this getter into the header file, you won't need to repeat the method's name in the comment :) To make it more clearer this is just a getter, you could name it IsDecodingThreadActive().

rellla replied:
I'll drop that, because it's use no longer

fwolter commented on softhddevice.cpp:
```diff
@@ -453,6 +464,52 @@ void cSoftHdDevice::OnEnteringState(enum State state) {
+			m_pVideoStream->StartDecoder(new cVideoDecoder(m_pRender->HardwareQuirks())); // starts decoding thread
+			// Audio is init lazily (includes starting thread)
+			m_skipstream = false;
+			break;
```
You could move this to the leaving action of DETACHED. Then, the attached state can be removed, or?

rellla replied:
I think not, because we need a state where the plugin init at startup is done?!

fwolter replied:
Which code do you mean by plugin init exactly?

Can't the init be done in OnLeavingState(DETACHED)?

rellla replied:
Let me think...

https://github.com/rellla/vdr-plugin-softhddevice-drm-gles/pull/113/files#diff-324c042a2af7447ee936b59994e67c93f289d913e7be332acb4df5a558573a78L212

m_state is DETACHED per default. cSoftHdDevice::Start() is triggered either by VDR itself - or at startup via MakePrimaryDevice.

So we can put this code into OnLeavingState(DETACHED) and then every change out of DETACHED will trigger this.
What should be done, if we set DETACHED via svdrp and only want a dedicated ATTA to leave DETACHED? The code can still stay in OnLeavingState(DETACHED) but I guess we need some state to enter or at least to react on?

Correct me if i'm wrong.

fwolter replied:
Looks like some code needs to be executed when the plugin starts and on the other hand code that needs only executed when the plugin is atached. I think the latter should be executed when leaving the detached state, and the other in Start().

When the plugin starts, the run-once code is executed, and then, the attach event is sent, executing the leaving-detach-only code.

rellla replied:
> Looks like some code needs to be executed when the plugin starts and on the other hand code that needs only executed when the plugin is atached. I think the latter should be executed when leaving the detached state, and the other in Start().
> 
> When the plugin starts, the run-once code is executed, and then, the attach event is sent, executing the leaving-detach-only code.

Which code for the run-once at plugin start do you have in mind? The code which is minimal needed at plugin start is kept in cPluginSoftHdDevice::cSoftPluginHdDevice and is creating the config and the device itself.

All other things can/should be run at OnLeavingState(DETACHED) 

I like the idea of having the complement OnLeavingState(DETACHED) <-> OnEnteringState(DETACHED) but do not see any functions which should stay at another place (Start()). 
But imho we need to keep ATTACHED as a do-nothing-but-exist-to-enter-and-leave state.

fwolter replied:
Hm, I don't quite understand the purpose of the attached state. It has the same transitions as STOP and has no entering nor leaving code. Why don't you transition from DETACHED to STOP directly?

rellla replied:
Ok, that way? We have some unnecessary function calls in OnEnteringState(STOP), but they shouldn't hurt.

fwolter commented on softhddevice.cpp:
```diff
@@ -323,95 +292,137 @@ void cSoftHdDevice::HandlePause(void)
-	m_pRender->DecodingThreadResume();
-	m_pRender->DisplayThreadResume();
+	if (needsResume) {
+		m_pVideoStream->DecodingThreadResume();
```
Did you try to let them simply resume a second time when cancelling the threads? Shouldn't hurt if the mutex count is 2 excatly before it is destroyed.

rellla replied:
I did not try that. Same as above - i'm not satisfied with it either way.
What do you mean with mutex count 2? Unlocking an already unlocked mutex results in undefined behaviour according do docs.

fwolter replied:
You're right, I thought the mutex was backed by a sempahor, but that isn't the case, so lets keep it for now and maybe refactor it when implementing synchronized cQueue.

fwolter reviewed (commented):
Good work! I like the ratio of removed to added lines while adding a completly new feature!

Thanks for fixing the dependency between threads/videostream/videodecoder, that bothered me, too.

Looks way more cleaner how the plugin initilization is done with the new state machine, now!

Will you extend the state diagram?

fwolter commented on DEVELOPER.md:
```diff
@@ -94,25 +94,32 @@ This is the model of the state machine implemented in softhddevice.cpp.
-    [*] --> Stop: Initialize
+    [*] --> Detached: Initialize
+
+    Detached --> Stop: StopEvent
```
Shouldn't this be AttachEvent? Although it works technically, it semantically makes more sense to have a dedicated event, IMHO.
```suggestion
    Detached --> Stop: AttachEvent
```

rellla replied:
Because we don't have ATTACHED state anymore, i dropped the AttachEvent also. We could keep the Event and do the same as with a StopEvent?

fwolter replied:
I think that'd be better and contributes to a better understanding of the code.

fwolter commented on softhddevice.cpp:
```diff
@@ -323,20 +293,40 @@ void cSoftHdDevice::HandlePause(void)
-void cSoftHdDevice::OnEventReceived(const Event& event) {
+void cSoftHdDevice::OnEventReceived(const Event& event)
+{
+	m_mutex.lock();
```
You could use a lock guard, then unlock() is automatically called when the scope is left (e.g. during an early return). In this case it doesn't matter, but the docs of std::mutex.unlock() have a note that this method is not recommended to be used directly, but for example with a lock guard.
```suggestion
	std::lock_guard<std::mutex> lock(m_mutex);
```

rellla replied:
done

rellla committed:
lock virtual methods against the state machine and use lock_guard

The device and it's methods are still callable in detached mode, but we don't have
cVideoRender, cVideoDecoder .....

Return early in case any device method is called in detached state. We need to lock access
to these methods - especially m_skipstream - because we access them from different threads.

Also lock methods that are called via softosd: OsdClose(), OsdDrawARGB(), GetOsdSize()

rellla commented:
Huh, that is buggy.... We need to keep the plugin/device interface open for vdr to communicate but need to avoid calls to methods os nullptr objects.
It's reproduceable with a channel change try in detached mode for example...

rellla commented:
Would it be suitable to begin every virtual device method (which accesses audio, render, decoder ...) with a lock guard and a check against m_skipStream?

fwolter commented:
Yes, why not. Did you play with CanReplay()? Maybe VDR doesn't call any method when it returns false?

rellla commented:
> Yes, why not. Did you play with CanReplay()? Maybe VDR doesn't call any method when it returns false?

Because the heavy locking could be expensive? CanReplay is a good catch, will try that. That would be the simpler variant.

rellla commented:
Any objections against doing some squashing here and fix the remaining issues?

fwolter commented:
Don't bother about the complexity of mutex locking/unlocking. Even a RPI2 would manage it with a tired smile.

> Any objections against doing some squashing here and fix the remaining issues?

Go ahead!

rellla commented:
> Don't bother about the complexity of mutex locking/unlocking. Even a RPI2 would manage it with a tired smile.

It's more about locking VDR actions against eachother, but this is probably no problem.

rellla committed:
stop replay before detach

and prevent channel switch in detached state

Because we StopReplay() before detach, STOP is always the last state.
We enter STOP again when leaving DETACHED.

rellla committed:
add dummy osd and pixmap for skins

rellla commented:
So, next round. It's working not that bad afaics...

rellla commented:
Another small cleanup and rebase. Because we do StopReplay() before entering DETACHED, we don't need to keep the last state...

fwolter commented on softhddevice.cpp:
```diff
@@ -290,16 +260,38 @@ cSpuDecoder *cSoftHdDevice::GetSpuDecoder(void)
+ */
+bool cSoftHdDevice::Replaying(void) const
+{
+	return cDevice::Replaying();
```
Did you leave these intentionally?

rellla replied:
Is there a better way to access cDevice in cPluginSoftHdDevice instead of wrapping both into cSoftHdDevice?
We need them here https://github.com/rellla/vdr-plugin-softhddevice-drm-gles/pull/113/files#diff-6a83bdc9eb73021ed778e16c499b0d32271c410b23c188f7fad3c5b5a9a0296bR291

fwolter replied:
If an inherited class does not override a method of the base class, the base class' method is called. If I see correctly, you can simply remove them (and the declaration in softhddevice.h).

fwolter commented on softhddevice.cpp:
```diff
@@ -504,10 +586,10 @@ void cSoftHdDevice::SetState(enum State newState)
-		LOGDEBUG("device: Changing state %s -> %s", StateToString(m_state), StateToString(newState));
+		LOGDEBUG("device: Preparing to enter state %s", StateToString(newState));
+		OnEnteringState(newState);
+		LOGDEBUG("device: State changed from %s -> %s", StateToString(m_state), StateToString(newState));
 		m_state = newState;
-		OnEnteringState(m_state);
-		LOGDEBUG("device: State changed to %s", StateToString(m_state));
```
Is this change still necessary, now that everything is protected by the mutex? A disadvantage of this change is that SetState() cannot be used in OnEnteringState() anymore, because it'd be overwritten immediately afterwards. We don't need that functionality at the moment, just a precaution...

rellla replied:
Probably not, because everything should be locked now. I'll revert it.

fwolter commented on softhddevice.cpp:
```diff
@@ -1389,3 +1533,38 @@ int cSoftHdDevice::PlayVideoPkts(AVPacket * pkt)
+{
+	std::lock_guard<std::mutex> lock(m_mutex);
+	return m_skipstream;
+}
```
Is this still a problem, now that we have the mutex? `m_skipstream` is set in OnEnteringState(), and if you revert the order in SetState(), it has exactly the same function as setting and reading m_skipstream, or do I miss something?

Just a minor improvement, if you use IsDetached() everywhere instead of m_skipstream directly, you won't need to lock the mutex in every method.

rellla replied:
I think, we need some function to indicate the detached status, because atm we ask for it in softhddevide-drm-gles.cpp and softhdosd.cpp.

But your suggestion is good, i'll try it

rellla replied:
And I dropped m_skipstream again and use m_state == DETACHED directly. Should be safe now.

fwolter commented on softhddevice.h:
```diff
@@ -203,21 +216,29 @@ class cSoftHdDevice:public cDevice
 	int m_videoAudioDelay;           ///< audio/video delay set via setup menu
 	bool m_grabActive;               ///< simple lock variable
 	                                 ///< skips a new grab request if the last one is still active
+	bool m_skipstream;               ///< skips audio and video data (in detached state)
```
This should be declared as `volatile`, otherwise, the state change might not be seen by other threads immediately after the change, leading to race conditions.

rellla replied:
Ok. I'm not awareof all the keywords...

rellla replied:
not needed anymore?

fwolter replied:
Just for completeness, I just learned that volatile in C++ does not make variables thread-safe. You can use `std::atomic<bool>` if it is not already protected by a mutex.

fwolter commented on softhddevice.h:
```diff
@@ -203,21 +216,29 @@ class cSoftHdDevice:public cDevice
 	bool m_grabActive;               ///< simple lock variable
 	                                 ///< skips a new grab request if the last one is still active
+	bool m_skipstream;               ///< skips audio and video data (in detached state)
+	mutable std::mutex m_mutex;      ///< mutex to lock the state machine
```
I needed to lookup what the `mutable` keyword does. It has something to do with const objects. Did you add this intentionally?

rellla replied:
Yes. Otherwise it won't build. We cannot change variables in a const function except they get the mutable attribute.

fwolter replied:
Ah, ok, I didn't consider this!

rellla commented on softhddevice.cpp:
```diff
@@ -1540,14 +1532,10 @@ void cSoftHdDevice::Attach(void)
- * Use m_skipstream here instead of State::DETACHED or State::SUSPENDED,
- * because that one is set right before entering and after leaving DETACHED/SUSPENDED.
- * The state change is done somewhere in between and we can't rely on that.
+ * Returns true, id the device is detached
```
Typo... will fix it.

rellla commented on softhddevice.cpp:
```diff
@@ -574,10 +574,10 @@ void cSoftHdDevice::SetState(enum State newState)
-		LOGDEBUG("device: State changed from %s -> %s", StateToString(m_state), StateToString(newState));
+		LOGDEBUG("device: Changing state %s -> %s", StateToString(m_state), StateToString(newState));
 		m_state = newState;
+		OnEnteringState(newState);
```
I'll be using m_state here again.

rellla committed:
readme: minor update

rellla committed:
version 1.2.0

fwolter commented on softhddevice.cpp:
```diff
@@ -290,16 +259,29 @@ cSpuDecoder *cSoftHdDevice::GetSpuDecoder(void)
+
+	LOGDEBUG("device: %s: %d", __FUNCTION__, canReplay);
+
+	return canReplay;
```
```suggestion
	bool canReplay = !IsDetached();

	LOGDEBUG("device: %s: %d", __FUNCTION__, canReplay);

	return canReplay;
```

fwolter reviewed (approved):
LGTM!
